### PR TITLE
Game controller

### PR DESCRIPTION
**Pull request by ellehallal**

### CODE REVIEW

File: __tests__/game_controller_test.js [hunk 1]
```diff
@@ -7,7 +7,101 @@ describe('GameController', () => {
...
+  });
+
+  it('allows a player to make a move', () => {
+    gameController.takeTurn(3)
+    expect(gameController.game.grid).toEqual([1, 2, 'x', 4, 5, 6, 7, 8, 9]);
+  });
+
+  it('allows a second player to make a move', () => {
+    gameController.takeTurn(3)
+    gameController.togglePlayer()
```
Comment (pelensky) [inline __tests__/game_controller_test.js]: Could the takeTurn method also toggle the player?

File: __tests__/game_controller_test.js [hunk 2]
```diff
@@ -7,7 +7,101 @@ describe('GameController', () => {
...
+    gameController.takeTurn(5)
+    expect(gameController.game.grid).toEqual([1, 2, 'x', 4, 'o', 6, 7, 8, 9]);
+  });
+
+  it('returns "Tie" when neither player has won after 9 moves', () => {
```
Comment (pelensky) [inline __tests__/game_controller_test.js:28]: Sorry to sound like a broken record, but dependency injection could avoid the need to do this! 

So if the Board constructor takes a grid, the Game takes a board, and the Game Controller takes a Game, we could make this a bit smaller! 

So

board.js
```
class Board
constructor(grid) {
    this.grid = grid;
    this.moves = 0;
    this.winningCombos = [
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [0, 3, 6],
      [1, 4, 7],
      [2, 5, 8],
      [2, 4, 6],
      [0, 4, 8]
    ];
    this.checkCombos = false;
  }
```

game.js
```
class Game {
  constructor(board) {
    this.board = board;
    this.player = new Player;
    this.grid = this.board.grid;
    this.player1 = this.player.player1;
    this.player2 = this.player.player2;
    this.currentPlayer = this.player1;
  }
```

GameController.js
```
class GameController {
  constructor(game) {
    this.game = game;
  }
```

So to do this exact same test, we could write it as 

```
it('returns "Tie" when neither player has won after 9 moves', () => {
const board = new Board(['o', 'x', 'x', 'x', 'o', 'o', 'o', 'x', 'x']);
const game = new Game(board);
const gameConstructor = new GameConstructor(game);

expect(gameControlller.isATieOrWon()).to Equal("Tie");
}) 
```

File: __tests__/game_controller_test.js [hunk 3]
```diff
@@ -7,7 +7,101 @@ describe('GameController', () => {
...
+    gameController.togglePlayer()
+    gameController.takeTurn(4)
+    gameController.togglePlayer()
+    gameController.takeTurn(6)
+    gameController.togglePlayer()
+    gameController.takeTurn(9)
+    expect(gameController.isATieOrWon()).toEqual("Tie");
+  });
+
+  it('returns "Win" when a player has won', () => {
```
Comment (pelensky) [inline __tests__/game_controller_test.js]: For these other ones you'd then just need to change the state of the board in the Board class!

File: src/board.js [hunk 4]
```diff
@@ -15,15 +15,15 @@ class Board {
     this.checkCombos = false;
   }
 
+  isValidPosition(position) {
```
Comment (pelensky) [inline src/board.js:18]: What happens if someone tries to place an x or o on a space that's already taken?

File: webpack.config.js [hunk 5]
```diff
@@ -1,10 +1,10 @@
 const path = require('path');
 
 module.exports = {
-  entry: './src/index.js',
+  entry: './src/index.js', // this defines the file which acts as the entry point between our logic files and webpage (minimises so that it loads faster)
```
Comment (pelensky) [inline webpack.config.js:4]: 👏

**Review by pelensky — APPROVED** [review]
LOOKS AMAZING!!!

File: src/board.js (hunk 4 above)
Comment (sohbaker) [reply src/board.js:18]: Thanks for flagging, we had this in the more complicated version of our start method but didn't include it in the refactored one! We've now added an if...else to handle that possibility.

File: __tests__/game_controller_test.js (hunk 1 above)
Comment (sohbaker) [reply __tests__/game_controller_test.js]: At the moment if takeTurn also toggles player, the game can't track whether that player has won after they've made their move.

File: webpack.config.js (hunk 5 above)
Comment (sohbaker) [reply webpack.config.js:4]: 😃

File: __tests__/game_controller_test.js (hunk 2 above)
Comment (sohbaker) [reply __tests__/game_controller_test.js:28]: Thanks Dan. This was super helpful and we've refactored our classes today to include dependency injection

**Pull request merged**